### PR TITLE
Add rocksdb to the benchmarks

### PR DESCRIPTION
**Pull request by tvaleev**
#103

### CODE REVIEW

**Review by mjp41 — COMMENTED** [review]
Thanks for looking into this.  I have a couple of questions

File: build-bench-env.sh [hunk 1]
```diff
@@ -663,6 +670,21 @@ if test "$setup_redis" = "1"; then
   popd
 fi
 
+if test "$setup_rocksdb" = "1"; then
+  phase "build rocksdb $version_rocksdb"
+
+  pushd $devdir
+  if test -d "rocksdb"; then
```
Comment (jvoisin) [inline build-bench-env.sh:677]: You can replace this with the `checkout` function :)

**Review by jvoisin — CHANGES_REQUESTED** [review]

File: build-bench-env.sh [hunk 2]
```diff
@@ -663,6 +670,21 @@ if test "$setup_redis" = "1"; then
...
+    echo "$devdir/rocksdb already exists; no need to download it"
+  else
+    git clone https://github.com/facebook/rocksdb.git
+  fi
+  cd rocksdb
+  git checkout "$version_rocksdb"
+  DISABLE_JEMALLOC=1 make all -j $procs
```
Comment (jvoisin) [inline build-bench-env.sh:684]: Don't build the whole project, only what's required :)

File: bench.sh [hunk 3]
```diff
@@ -23,14 +23,14 @@ alloc_libs="sys="      # mapping from allocator to its .so as "<allocator>=<sofi
...
-readonly tests_all4="z3 spec spec-bench"
+readonly tests_all4="z3 spec spec-bench rocksdb"
 
 readonly tests_all="$tests_all1 $tests_all2 $tests_all3 $tests_all4"
 readonly tests_allt="$tests_all1 $tests_all2"  # run with 'allt' command option
 
 tests_run=""
 tests_exclude=""
-readonly tests_exclude_macos="sh6bench sh8bench redis"
+readonly tests_exclude_macos="sh6bench sh8bench redis rocksdb"
```
[the remark this answers is not in the document]
Comment (tvaleev) [reply bench.sh:33]: I cannot test on OSX :) But if it ok I can delete rocksdb from tests_exclude_macos.

File: build-bench-env.sh (hunk 2 above)
Comment (tvaleev) [reply build-bench-env.sh]: Done

File: build-bench-env.sh (hunk 1 above)
Comment (tvaleev) [reply build-bench-env.sh]: Done

File: bench.sh [hunk 4]
```diff
@@ -23,7 +23,7 @@ alloc_libs="sys="      # mapping from allocator to its .so as "<allocator>=<sofi
 readonly tests_all1="cfrac espresso barnes redis lean larson-sized mstress rptest gs"
 readonly tests_all2="alloc-test sh6bench sh8bench xmalloc-test cscratch glibc-simple glibc-thread"
 readonly tests_all3="larson lean-mathlib malloc-large mleak rbstress cthrash"
-readonly tests_all4="z3 spec spec-bench"
```
Comment (mjp41) [inline bench.sh:26]: Could you add it to `tests_all2` so we get some runs in CI?  At the moment, CI is only testing building it, and not running.  We might not want it in all CI configs if it takes too long, but I would like to see it before we merge.

File: bench.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (jvoisin) [reply bench.sh]: That's how everyone is debugging OSX :D

File: bench.sh (hunk 4 above)
Comment (tvaleev) [reply bench.sh:26]: Done

**Comment by mjp41** [comment]
Alpine is failing with
```
run 1: rocksdb sn-sec: LD_PRELOAD=/__w/mimalloc-bench/mimalloc-bench/extern/snmalloc/release/libsnmallocshim-checks.so /__w/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench --benchmarks=fillrandom
Please install gflags to run rocksdb tools
Command exited with non-zero status 1
rocksdb     sn-sec 0m 0.00s 25632 0.00 0.00 0 355
```
I think we need to install stuff for Alpine too, and not just Ubuntu.  Mac is failing similarly.
```
---- 1: rocksdb
run 1: rocksdb sys: SYSMALLOC=1 /Users/runner/work/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench --benchmarks=fillrandom
Please install gflags to run rocksdb tools
Error: Process completed with exit code 1.
```

Fedora is failing to build. 
```
/__w/mimalloc-bench/mimalloc-bench/extern /__w/mimalloc-bench/mimalloc-bench
HEAD is now at 8e0f49525 Update version
make: which: No such file or directory
$DEBUG_LEVEL is 1
make: which: No such file or directory
make: -c: No such file or directory
make: -c: No such file or directory
make: -c: No such file or directory
make: -c: No such file or directory
make: -c: No such file or directory
Makefile:173: Warning: Compiling in debug mode. Don't use the resulting binary in production
make: -c: No such file or directory
make: -c: No such file or directory
```
which makes no sense to me.  Looks like it can't find `which`?

**Comment by mjp41** [comment]
So Mac OS X seems to fail with missing snappy?
```
run 1: rocksdb sys: SYSMALLOC=1 /Users/runner/work/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench --benchmarks=fillrandom
open error: Invalid argument: Compression type Snappy is not linked with the binary.
Error: Process completed with exit code 1.
```
And Alpine:
```
run 1: rocksdb tc: LD_PRELOAD=/__w/mimalloc-bench/mimalloc-bench/extern/tc/.libs/libtcmalloc_minimal.so /__w/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench --benchmarks=fillrandom
open error: Invalid argument: Compression type Snappy is not linked with the binary.
Command exited with non-zero status 1
rocksdb     tc    0m 0.00s 50304 0.00 0.00 0 1183
```


Hoard is crashing on Ubuntu on RocksDB:
```
run 1: rocksdb hd: LD_PRELOAD=/home/runner/work/mimalloc-bench/mimalloc-bench/extern/Hoard/src/libhoard.so /home/runner/work/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench --benchmarks=fillrandom
... finished 1000000 ops                              
RocksDB:    version 7.3
Date:       Sun Jun 19 08:51:03 2022
CPU:        2 * Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
CPUCache:   51200 KB
Received signal 11 (Segmentation fault)
#0   /home/runner/work/mimalloc-bench/mimalloc-bench/extern/Hoard/src/libhoard.so(customfree+0xa7) [0x7fe4e40b7527] ??	??:0	
#1   /home/runner/work/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench(+0x10bec7) [0x555c94db7ec7] ??	??:0	
#2   /home/runner/work/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench(+0xd028b) [0x555c94d7c28b] ??	??:0	
#3   /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7fe4e3a2a083] ??	??:0	
#4   /home/runner/work/mimalloc-bench/mimalloc-bench/extern/rocksdb/db_bench(+0xcf4be) [0x555c94d7b4be] ??	??:0	
```

I'll try to replicate this locally tomorrow, and we should perhaps raise an issue against Hoard.

**Comment by tvaleev** [comment]
> I'll try to replicate this locally tomorrow, and we should perhaps raise an issue against Hoard.

Strange...Cannot reproduce locally on my Ubuntu Ubuntu 20.04.4 LTS.

**Comment by tvaleev** [comment]
I've got for Fedora:
```
/usr/bin/ld: final link failed: No space left on device
collect2: error: ld returned 1 exit status
make: *** [Makefile:1601: block_based_table_reader_test] Error 1
make: *** Waiting for unfinished jobs....
Error: Process completed with exit code 2.

```
How to avoid this? :)

**Comment by jvoisin** [comment]
`No space left on device` we should remove intermediary object files in the CI I think, likely in another PR :/

**Comment by jvoisin** [comment]
#140 might help

**Comment by jvoisin** [comment]
@tvaleev can you rebase on the current master?

**Comment by tvaleev** [comment]
> @tvaleev can you rebase on the current master? The disk space issue should be gone now :)

Looks like the same :( 
Just an idea...Looks like is possible to delete mimalloc-bench/extern/redis-6.2.7.tar.gz after untar.

**Comment by tvaleev** [comment]
Again `final link failed: No space left on device` on Fedora.

**Comment by mjp41** [comment]
If you build rocksdb first on Fedora will it work?  Could we tidy up the intermediates for rocksdb, and then build everything else?

**Comment by tvaleev** [comment]
> If you build rocksdb first on Fedora will it work? Could we tidy up the intermediates for rocksdb, and then build everything else?

I tried to move the build of rocksdb ahead all other benchmarks but again the same failed result :( Next I can only try to move it ahead of allocators...

**Comment by jvoisin** [comment]
You could try to download an archive of the latest version, instead of the whole git repo, extract it, remove it, compile rocksdb, remove all the object files.

This should work :)

**Comment by tvaleev** [comment]
> You could try to download an archive of the latest version, instead of the whole git repo, extract it, remove it, compile rocksdb, remove all the object files.
> 
> This should work :)

Done :)

File: build-bench-env.sh [hunk 5]
```diff
@@ -640,12 +640,24 @@ fi
...
+    rm "./rocksdb-$version_rocksdb.tar.gz"
+  fi
+
+  cd "rocksdb-$version_rocksdb"
   if grep -q 'ID=fedora' /etc/os-release 2>/dev/null; then
     DISABLE_WARNING_AS_ERROR=1 USE_JEMALLOC=no MALLOC=libc BUILD_TLS=no make -j $procs
   else
     DISABLE_JEMALLOC=1 make db_bench -j $procs
   fi
+  find . -name "*.o" -exec rm -f {} \;
```
Comment (jvoisin) [inline build-bench-env.sh:684]: `find . -name "*.o" -delete` is better :)

**Comment by jvoisin** [comment]
I usually disable things in the .github/workflows/all.yaml file directly

**Pull request merged**